internal/store/certificatesigningrequest.go

```diff
@@ -31,20 +31,40 @@ import (
 )
 
 var (
+	descCSRAnnotationsName     = "kube_certificatesigningrequest_annotations"
+	descCSRAnnotationsHelp     = "Kubernetes annotations converted to Prometheus labels."
 	descCSRLabelsName          = "kube_certificatesigningrequest_labels"
 	descCSRLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
 	descCSRLabelsDefaultLabels = []string{"certificatesigningrequest", "signer_name"}
 )
 
-func csrMetricFamilies(allowLabelsList []string) []generator.FamilyGenerator {
+func csrMetricFamilies(allowAnnotationsList, allowLabelsList []string) []generator.FamilyGenerator {
 	return []generator.FamilyGenerator{
+		*generator.NewFamilyGenerator(
```

Comment:
As asked in https://github.com/kubernetes/kube-state-metrics/pull/1495#issuecomment-896964375 I'd suggest to add `MetricFamily` only when `allowAnnotationsList` is provided for this k8s Kind. For example for `csr` it would be: https://github.com/kubernetes/kube-state-metrics/pull/1495/files?file-filters%5B%5D=.go#diff-ee6d72cabf35aed3a2691f1505a707f49a38e752478c98a66d1dfc5ec6b16e51R90
Consider starting ksm with:
`--metric-labels-allowlist=pods=[team] --metric-annotations-allowlist=namespaces=[channel]`
Which would create following metrics for each k8s Kind:
```
kube_<kind>_labels{namespace="ns",<kind>="name"} 1
kube_<kind>_annotations{namespace="ns",<kind>="name"} 1
```
On small testing cluster I'm getting:
```
$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | head
kube_daemonset_annotations{namespace="kube-system",daemonset="local-volume-provisioner"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="kube-router"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="node-exporter"} 1
kube_daemonset_annotations{namespace="ceph",daemonset="csi-rbdplugin"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="filebeat"} 1
kube_daemonset_labels{namespace="ceph",daemonset="csi-rbdplugin"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="filebeat"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="local-volume-provisioner"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="kube-router"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="node-exporter"} 1

$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | wc -l
803
```
These metrics have no sense and only lead to increase of cardinality. From cli args I only interested in `kube_pod_labels` and `kube_namespace_annotations`.
Alternative implementation from https://github.com/kubernetes/kube-state-metrics/pull/1495 would yield:
```
$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | head
kube_namespace_annotations{namespace="vault"} 1
kube_namespace_annotations{namespace="cert-manager"} 1
kube_namespace_annotations{namespace="default",annotation_channel="@UQQ184EHF"} 1
kube_namespace_annotations{namespace="kube-node-lease"} 1
kube_namespace_annotations{namespace="kube-system",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="it",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="ceph"} 1
kube_namespace_annotations{namespace="kata-agents",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="kuberos"} 1
kube_namespace_annotations{namespace="kube-public"} 1

$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | wc -l
90  #all pods returns labels
```
Which is more in line with `--metric-allowlist` usage.
In case starting ksm without `--metric-labels-allowlist/--metric-annotations-allowlist` cli arguments set, current implementation would only increase unneeded cardinality. New metrics should be disabled by default https://github.com/kubernetes/kube-state-metrics/issues/941#issuecomment-729560572
> By default this would be all disabled and users could just enable annotations they are interested in.

Reply:
I hear you but I made this PR so that it replicates the behaviour of `kube_<kind>_labels` metrics.

What you want makes sense to me but I think that it probably should be a separate PR.

@lilic, @dgrisonnet, care to comment about this ?

Reply:
I think it would make sense to hide the `labels` and `annotations` metrics if the allowlists are empty :+1:. Having it changed in a separate PR would make more sense since this effort is already huge and it will only make things harder to review.